### ISSUE TITLE
Bug when using :: to type variables

### ISSUE THREAD

**Issue by dwightguth** [kframework/k-legacy#4]
module TEST

rule A:KItem ::Int ~> A ::Int => .

endmodule

This should compile no problem, but instead it gives the following error:

[Error] Critical: Variable 'A' declared with two different sorts: Int and KItem
        File: /home/dguth2/ktest/test.k
        Location: (3,23,3,24)
        Compilation Phase: Variable type inference

I need this fixed right away so I can tag a release of the K framework that the Python semantics works with.

**Comment by grosu** [comment]
Since Dwight will leave our group in a few days and we really want to have
the Python semantics work, this is sort of urgent.

Thanks,
Grigore

**Comment by radumereuta** [comment]
Hmm... this is tricky.
This isn't how I implemented the variable type inferencer/disambiguator.
When you write A::Int, I'm considering the type of variable A as having the type Int everywhere in the rule, but typed syntactically, which will not generate the runtime check.

Since you declare it at first with KItem, and then with Int, I'm reporting an error.
If you want to avoid A to be typed as Int, then use parenthesis: (A)::Int

Somehow I feel like this is not the end of this discussion, but this is how we decided to do it a long time ago.

Andrei S. also encountered this problem last week and he accepted my workaround.

**Comment by dwightguth** [comment]
I will add parentheses for now so I can make it work but I am not at all happy with this solution. If I type a variable with two conflicting types it should not report an error. It should add a runtime check for every : type and, type the variable statically only locally. So:

rule A:KItem ::Int ~> A::Int ~> A

should type the first as KItem the second as Int and the third as K. If you want you can produce a hidden warning in this case. But it should not be an error.

**Comment by dwightguth** [comment]
I turned this into an enhancement because according to Radu's statement it's not exactly a bug. And I can make python compile now, so it's no longer priority critical. But we should leave this open and you guys should discuss it in the next K meeting, assuming I'm not the only one who thinks my proposal is better than the current behavior.

**Comment by dlucanu** [comment]
I find a bit confusing the use of the same variable with different sorts in the same rule. Up to now we had the convention that a variable has a unique sort per rule and this sort can be declared just once  for the cases when cannot be inferred.
If we adopt this change,  the definition for scoping becomes very tricky and hard to use it (I do not see yet a simple one). Moreover, all existing definition must be revised in order to specify each time the type of the variable when this is not K.
The current solution is cleaner form the user's perspective.

**Comment by dwightguth** [comment]
I don't think it's at all confusing to imagine that if a global sort cannot be assigned for a variable in a rule then the global sort is K except when explicitly locally declared otherwise. Also, variable sort is already nonlocal because a variable without an explicitly declared sort can have any sort depending on what else is in the rule. So I don't see how we are making anything more confusing. Finally, no changes to definitions will need to be made because we will be making a corner case which currently causes an error to work. So it will have no impact whatsoever on the semantics of the programs which we currently allow. It will only affect the semantics of programs which currently we give an error on.

The reason why I propose this change is because the behavior we exhibited in my original post is neither clean nor particularly obvious. Rules similar to the one I described in my post exist in the Python semantics and are necessary in order to be able to unambiguously parse a very complex grammar. It is very intuitive to imagine a rule in which a term is syntactically assumed to have sort Int but at runtime we only enforce that it has sort KItem. The rule I wrote above is the obvious and natural way one would describe such a definition and the fact that it does not work is very confusing, as is obvious from the fact that I viewed it as a bug. Furthermore, the workaround Radu proposes is acceptable for now but simply not remotely suitable as a permanent solution because it gives an undesirable semantics to the bracket operator. It is not at all clear to anyone why A:Exp and (A):Exp should have such different semantics, and as far as I am aware there has not been a single person who has deduced that that was the correct workaround without being told by Radu, which indicates very strongly that it is counterintuitive and likely to confuse users.

In summary, I find it disingenuous to claim that the existing semantics is desirable for users when I am a user and I claim otherwise.

**Comment by denis-bogdanas** [comment]
Ok Guys,
I've read the whole topic and am terribly confused. I was only able to infer two things:
1. The construct :: now have a meaning in K language, A::T denotes the type of the variable A, but is somehow different from A:T.
2. In the same rule, a variable A may have different types in different places.

Can someone please explain me what all of this means? Or at least point me to the right topic elsewhere. I have two problems with all this novelties:
1. The construct A::T is widely used in Java Semantics, replacing this with something else would look horrible.
2. Why would anyone assign different types in different places to the same var, all of this in one single rule?

**Comment by grosu** [comment]
This is all I remember from our past discussions:

":" means "hard sorting", that is, the term must have the claimed sort at runtime; a runtime check will be performed.

"::" means "casting", that is, the term is assumed to have the claimed sort for static purposes, to parse the rule, but no check is done at runtime.

"(" and ")" can be used to delimit the term, because both ":" and "::" have maximum priority in terms of parsing.

Currently, a runtime check matters in terms in performance, because the sort membership predicate takes linear time in the size of the checked term.

Therefore, if you can avoid using ":", then in terms of speed you are better off not using it.  However, you sometimes cannot avoid id, for example:

... X::Id = V:Val; ...

The above may appear in the semantics of assignment.  X::Id tells the parser what it needs to parse the assignment, and since you know that you will never in your semantics change the lhs of an assignment and the original program only had program variables in the lhs of assignment, you do not need to worry about writing X:Id, which would be more expensive.  You are telling the parser "trust me and shut up".  However, V:Val is needed, because you want the rule to only apply when V is a value.  Writing V::Val instead of V:Val would make the parser happy, but the assignment rule may apply before V is evaluated to a value.

Other than the above, I remember there were all kinds of discussions on how to limit the above due to SDF constraints that Radu sensed.  Basically, Radu said that certain things we proposed are impossible to achieve with SDF without a huge performance price to pay.  At the time, since we were under time pressure we decided to move on.  But I do not remember to have unanimously agreed with anything.  Since this is mostly a front-end issue, which we will redesign anyway once we have Kore in place and running, I propose that we postpone (again) this discussion for then.

Grigore

**Comment by dwightguth** [comment]
Denis: Consider the following rules in Python:

rule : E : E2 => (ref("slice") (ref("None"), E ::Exp, E ::Exp)) [macro]
rule (. => bind(X:Id, ref(B:ObjId))) ~> ((X)::Target = Ts:AssignTargets => Ts) = obj(B,_)

In the first of these rules, a slice expression is converted into a call to the "slice" function. Because we need to add the .Exps to the call, we need E to be of sort Exp. However, because macros may already have applied, E may in fact not have the exact type Exp at runtime. So we use ::Exp to indicate that statically E has the sort Exp, and therefore we should add a .Exps, but that no runtime check should be made.

In the second of these rules, an assignment to a variable is converted into a binding in the environment. In python, if "x" and "y" are variables, then "x = y" can refer to four distinct pieces of syntax. It can refer to a parameter to a function declaration x with default value y, it can refer to a keyword argument x to a function call whose value is y, it can refer to the assignment of the value of y to the variable x, or it can refer to a list of assignment targets as it does in the subexpression "x = y" of the expression "x = y = z". Because these constructs have different KLabels, it is necessary at parse time to distinguish between them. In the second rule above, we assert that X is a Target and Ts is an AssignTargets in order to select the last of these four constructs syntactically. However, not all Targets should cause this rule to apply. In fact only if X is literally an Id should the rule apply. However, if instead of (X)::Target we said X:Id then it would be ambiguous because several other of these constructs can have an Id on the left-hand side. Thus we must give X one sort semantically and another sort syntactically.

Does this make sense? I'm sorry about the syntax conflict with your semantics but it shouldn't be too hard to change using sed. You just use something like "s/::([^=])/:::\1/g". It might look less nice, but in the long run we will have the ability to use custom parsers for K definitions, and you should be able to change the syntax of K so that something other than :: is used for syntactic casts. You just need to bear up a little while longer.

**Comment by denis-bogdanas** [comment]
Now I understand the arguments, but they don't look convincing. Overall, the construct :: looks like a cheat to me. 

Grigore: The argument of performance is very unconvincing. An execution engine that is millions of times slower than a production compiler will hardly be more appealing to use if it is a few times faster at the expense of less runtime checks. Since K is a language primarily targeted for prototyping rather than production, having stronger checks looks like a more valuable quality that faster execution.

Dwight's example 1: You have a construct like

K ::= "foo" "(" Id ")"

but then you want a rule like
rule foo(X::Id) => Y
to match a term like foo(Exp)
Aren't you cheating your own syntax? Why not defining "foo" "(" Exp ")" then?

However, I don't know what macros do, so I might be unaware of the whole picture.

Dwight's example 2:
rule (. => bind(X:Id, ref(B:ObjId))) ~> ((X)::Target = Ts:AssignTargets => Ts) = obj(B,_)

Dwight says that it may have four different meanings, represented by four KLabel-s, and the casting (X)::Target have the role to match just one case. Wouldn't it be more straightforward to express this rule in the labeled form?

As a final thought, by adding the construct "::", K ceases to be a strongly typed language. By strongly typed I mean if we have "foo" "(" Id ")" and no other foo syntaxes, then never ever the expression inside foo could be non-Id. Strong typing was never enforced in practice anyway, but at least I assumed this is a development goal.

**Comment by grosu** [comment]
No, strong typing was never a development goal, and it will never be, because it is a poor idea for programming language semantics, which limits the way you can do things.  Perhaps SOS is the best example, followed by reduction semantics with evaluation contexts.  CHAM also decided to free itself from syntax, and if you read their papers they bring good reasons there, too.  And there is no cheating.  It was all planned.  K is NOT rigid to syntax, by design.  And K will be fast one day, fast enough that K definitions will compete with production interpreters.  If you read the messages on the list, it already does when it gets to type checking/inference.

Checking the membership of a term to a sort is a much more expensive operation than it may seem, maybe even impossible to compute in all cases.  For example, the sort of S1:Stmt ~> [];S2:Stmt should be Stmt, because its class of representatives (terms obtained by applying heating/cooling) in a functional language where ";" is strict(1) contains S1:Stmt; S2:Stmt, which parses as a Stmt.  Moreover, in an imperative language I may have a macro S1; S2 => S1 ~> S2, which, being a macro, applies everywhere, statically.  This way, you never have to worry about giving semantics to ";".  Now consider

  rule {S::Stmt} => S ...  (... stands for recovering the environment)

If you replace ::Stmt with :Stmt, then your rule will never apply for blocks with more than one statement, because the macro expands, for example, {x=3; y=4;} into {x=3; ~> y=4;}, and x=3; ~> y=4; parses as a K term but not as a Stmt.  You need the soft casting ::Stmt to convince the parser to accept the rule.

Grigore

**Comment by radumereuta** [comment]
Starting with aca2c65a33 this is fixed.
X::Exp is now a wrapper that does not restrict the internal sort.
In this case, the variable X will be inferred of having sort K
